File: addBinary.py
class Solution:
    def addBinary(self, a, b):
        """
        :type a: str
        :type b: str
        :rtype: str
        """
        # max value among a & b        
        m = max(len(a), len(b))
        print("There are %s character in the binary number"%(m))
        
        # initialize result
        result = ''
        
        # fill zero into a or b
        a = a.zfill(m)
        b = b.zfill(m)
        print(a, b)
        
        # carry value in binary sum
        carry = 0 
        
        # adding value from right to left with range(start, stop, step) like how we would do it using pen and paper
        for i in range(m-1, -1, -1):
            t = carry
            t += 1 if a[i] == '1' else 0
            t += 1 if b[i] == '1' else 0
            carry = 0 if t < 2 else 1
            result = ('1' if t % 2 == 1 else '0') + result
        
        # add addtional digit to binary of carry > 0
        if carry != 0:
            result = '1' + result
        
        return result.zfill(m)
            
            
        
        
        
        
        

File: addBinary.swift
class Solution {
    func addBinary(_ a: String, _ b: String) -> String {
        let c1:Int = a.count
        let c2:Int = b.count
        let m:Int = [c1, c2].max()!
        var A = String(repeating: "0", count: m-c1) + a
        var B = String(repeating: "0", count: m-c2) + b
        var carry = 0
        var result = ""  
        for i in (0...m-1).reversed() {
            var total = carry
            let index = A.index(A.startIndex, offsetBy: i)
            print(A[index])
            print(B[index])
            if A[index] == "1" {
                total = total + 1
            } else {
                total = total + 0
            }
            if B[index] == "1" {
                total = total + 1
            } else {
                total = total + 0
            }
            if total < 2 { 
                carry = 0 
            } else {
                carry = 1
            }
            if total%2 == 1 { result = "1" + result } else { result = "0" + result}
        }
        // add additional 1 for larger value
        if carry != 0 { 
            result = "1" + result 
        } 
        return result
    }
}

